geochem_inverse_optimize.py

```diff
@@ -202,6 +208,38 @@ def solve(
 
         return downstream_preds, upstream_preds
 
+    def solve_montecarlo(
+        self,
+        observation_data: ElementData,
+        relative_error: float,
+        num_repeats: int,
+        regularization_strength: float,
+        solver: str = "gurobi",
+    ):
+        predictions_down_mc = {}
+        predictions_up_mc = {}
+        i = 1
+        while i < num_repeats:
+            observation_data_resampled = observation_data.copy()
+            observation_data_resampled.update(
+                (sample, value * np.random.normal(loc=1, scale=relative_error / 100))
+                for sample, value in observation_data_resampled.items()
+            )
+            element_pred_down, element_pred_upstream = self.solve(
+                observation_data=observation_data_resampled,
+                solver=solver,
+                regularization_strength=regularization_strength,
+            )  # Solve problem
+            for sample_name in element_pred_down.keys():
+                if i == 1:  # Initiate output array
```

Comment:
Use `defaultdict` to simplify the logic here.

Reply:
Great tip thanks!